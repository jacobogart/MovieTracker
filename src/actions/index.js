export const addDiscoverMovies = (movies) => ({
  type: 'ADD_DISCOVER_MOVIES',
  movies
});

export const updateUser = (id, name, favorites) => ({
  type: 'UPDATE_USER',
  id, 
  name,
  favorites
});

export const logoutUser = () => ({
  type: 'LOGOUT_USER'
});
