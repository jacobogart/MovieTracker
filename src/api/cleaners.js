import { imageUrl } from '../utils/API-logins';

export const cleanAllMovies = (moviesArray) => {
  const cleanMovies = moviesArray.map(movie => {
    const { 
      id, 
      title, 
      poster_path, 
      overview, 
      release_date, 
      vote_average 
    } = movie;

    return { 
      id, 
      title, 
      posterImg: `${imageUrl}${poster_path}`, 
      overview, 
      releaseDate: release_date, 
      rating: vote_average 
    };
  });

  return cleanMovies;
};

export const cleanPeople = (data) => {
  const director = data.crew.filter(person => person.job === "Director");
  const writer = data.crew.filter(person => person.job === "Writer");
  const cast = data.cast.map(person => {
    let { name, character, profile_path,  cast_id} = person;

    return { 
      name, 
      character, 
      headshot: `${imageUrl}${profile_path}`, 
      id: cast_id 
    };
  });

  return { director, writer, cast };
};