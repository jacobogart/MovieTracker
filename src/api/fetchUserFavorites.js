import { serverUrl } from "./pathNames";

export const fetchUserFavorites = (id) => {
  const url = `${serverUrl}/${id}/favorites`;

  return fetch(url)
    .then(response => {
      if (!response.ok) {
        throw Error('Failed to fetch user favorites');
      } else  {
        return response.json();
      }
    })
    .then(result => result.data);
};