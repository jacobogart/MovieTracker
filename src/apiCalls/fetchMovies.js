import { jacobKey, baseUrl } from '../utils/API-logins'

export const fetchMovies = () => {
  fetch(`${baseUrl}discover/movie${jacobKey}`)
    .then(response => response.json())
    .then(result => console.log(result));
}

