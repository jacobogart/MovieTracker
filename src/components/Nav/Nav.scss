.Nav {
  display: flex;
  justify-content: space-between;

  .main-nav-wrapper {

  }

  .nav-link {
    // width: 200px;
  }
  
  .account-wrapper {
    margin: 0;

    .login-link {
      display: inline-block;
      margin: 0;
    }
  }
}