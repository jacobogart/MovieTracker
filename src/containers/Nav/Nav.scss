.Nav {
  display: flex;
  justify-content: space-between;
  padding: 30px;

  .nav-left {
    display: flex;
    flex-direction: row;

    .header-logo {
      margin: 0 15px 0 0;
    }
  }

  .nav-right {
    display: flex;

    .login-link {
      margin: 0 10px;
      padding: 5px;

      &:hover {
        cursor: pointer;
      }
    }
  }

  .nav-link {
    margin: 0 10px;
    padding: 5px;
  }

  .login-dropdown {
    background: #999;
    padding: 20px;
    position: absolute;
    top: 70px;
    right: 30px;
    z-index: 1;
  }
}