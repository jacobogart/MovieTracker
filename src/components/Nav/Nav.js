import React, { Component } from 'react';
import { NavLink } from 'react-router-dom';
import LoginForm from '../../containers/LoginForm/LoginForm';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import { logoutUser } from '../../actions';

class Nav extends Component {
  constructor() {
    super();

    this.state = {
      showLogin: false
    }
  }

  toggleLogin = () => {
    this.setState({ showLogin: !this.state.showLogin });
  }

  handleLogout = () => {
    this.props.logoutUser();
  }

  render() {
    let accountNav;
    const { name } = this.props.user;

    if (!this.props.user.name) {
      accountNav =
        <div className="account-wrapper">
            <p 
              className="login-link nav-link"
              role="link"
              onClick={this.toggleLogin}
            >
              LOGIN
            </p>
            <NavLink to="/signup" className="nav-link">
              SIGN UP
            </NavLink>
          </div>
    } else {
      accountNav =
        <div className="account-wrapper">
          <p className="userGreeting">Hi, {name}!</p>
          <Link to="/" className="logout-link" onClick={this.handleLogout}>Logout</Link>
        </div>
    }

    return (
      <nav className="Nav">
        <div className="nav-left">
          <img src="" alt="Logo" />
          <div className="main-nav-wrapper">
            <NavLink to="/" className="nav-link">
              DISCOVER
            </NavLink>
            <NavLink to="/favorites" className="nav-link">
              FAVORITES
            </NavLink>
            <NavLink to="/watchlist" className="nav-link">
              WATCHLIST
            </NavLink>
          </div>
        </div>
        {accountNav}
        {this.state.showLogin && <LoginForm toggleLogin={this.toggleLogin} />}
      </nav>
    );
  }
}

const mapStateToProps = (state) => ({
  user: state.user
});

const mapDispatchToProps = (dispatch) => ({
  logoutUser: () => dispatch(logoutUser())
})

export default connect(mapStateToProps, mapDispatchToProps)(Nav);