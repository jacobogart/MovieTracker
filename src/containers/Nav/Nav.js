import React, { Component } from 'react';
import { NavLink } from 'react-router-dom';
import LoginForm from '../../containers/LoginForm/LoginForm';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import { logoutUser } from '../../actions';

class Nav extends Component {
  constructor() {
    super();

    this.state = {
      showLogin: false,
    }
  }

  toggleLogin = () => {
    this.setState({ showLogin: !this.state.showLogin });
  }

  hideLogin = () => {
    this.setState({ showLogin: false })
  }

  handleLogout = () => {
    this.props.logoutUser();
  }

  render() {
    const { name } = this.props.user;
    let accountNav;
    let loginActive = '';

    if (this.state.showLogin) {
      loginActive = 'active';
    }

    if (!this.props.user.name) {
      accountNav =
        <div className="nav-right">
            <a 
              href="#"
              className={`login-link nav-link ${loginActive}`}
              onClick={this.toggleLogin}>
              LOGIN
            </a>
            <NavLink to="/signup" className="nav-link" onClick={this.hideLogin}>
              SIGN UP
            </NavLink>
          </div>
    } else {
      accountNav =
        <div className="greeting-wrapper">
          <p className="user-greeting">HI, {name.toUpperCase()}!</p>
          <Link to="/" className="logout-link" onClick={this.handleLogout}>Logout</Link>
        </div>
    }

    const loginDropdown =
      <article className="login-dropdown">
        <LoginForm hideLogin={this.hideLogin} />
      </article>

    return (
      <nav className="Nav">
        <div className="nav-left">
          <Link to="/" className="main-heading"><h1>MOVIE TRACKER</h1></Link>
          <div className="main-nav-wrapper">
            <NavLink to="/" className="nav-link">
              EXPLORE
            </NavLink>
            <NavLink to="/favorites" className="nav-link">
              FAVORITES
            </NavLink>
            <NavLink to="/watchlist" className="nav-link">
              WATCHLIST
            </NavLink>
          </div>
        </div>
        {accountNav}
        {this.state.showLogin && loginDropdown}
      </nav>
    );
  }
}

const mapStateToProps = (state) => ({
  user: state.user
});

const mapDispatchToProps = (dispatch) => ({
  logoutUser: () => dispatch(logoutUser())
})

export default connect(mapStateToProps, mapDispatchToProps)(Nav);