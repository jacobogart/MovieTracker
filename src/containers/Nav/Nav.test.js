import React from 'react';
import { shallow } from 'enzyme';
import { Nav, mapStateToProps, mapDispatchToProps } from './Nav';
import { mockUser, mockCleanMovie, mockUncleanMovie } from '../../api/mockData';
import { fetchUserFavorites } from '../../api/fetchUserFavorites';
import { cleanAllMovies } from '../../api/cleaners';
import { logoutUser, addFavoriteMovies, addWatchlistMovies } from '../../actions';
import { fetchMovies } from '../../thunks/fetchMovies';


jest.mock('../../api/fetchUserFavorites');
jest.mock('../../api/cleaners');
jest.mock('../../thunks/fetchMovies');

fetchUserFavorites.mockImplementation(() => Promise.resolve([mockUncleanMovie]));
cleanAllMovies.mockImplementation(() => Promise.resolve([mockCleanMovie]));

const mockLogoutUser = jest.fn();
const mockResetMovies = jest.fn();
const mockDisplayFavorites = jest.fn();
const mockDisplayWatchlist = jest.fn();

describe('Nav', () => {
  let wrapper, instance;

  beforeEach(() => {
    wrapper = shallow(
      <Nav 
        user={mockUser}
        logoutUser={mockLogoutUser}
        resetMovies={mockResetMovies}
        displayFavorites={mockDisplayFavorites}
        displayWatchlist={mockDisplayWatchlist}
      />
    );
    instance = wrapper.instance();
  });

  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should have a detailed state', () => {
    const defaultState = { showLogin: false };
    expect(wrapper.state()).toEqual(defaultState);
  });

  describe('Event Listeners', () => {
    it("should invoke toggleLogin when login is clicked", () => {
      let wrapper = shallow(
        <Nav 
          user={{}}
          logoutUser={mockLogoutUser}
          resetMovies={mockResetMovies}
          displayFavorites={mockDisplayFavorites}
          displayWatchlist={mockDisplayWatchlist}
        />);
      wrapper.find('.login-link').simulate("click");
      expect(wrapper.state('showLogin')).toEqual(true);
    });

    it("should invoke toggleLogin when login is clicked", () => {
      let wrapper = shallow(
        <Nav
          user={{}}
          logoutUser={mockLogoutUser}
          resetMovies={mockResetMovies}
          displayFavorites={mockDisplayFavorites}
          displayWatchlist={mockDisplayWatchlist}
        />
      );
      wrapper.instance().setState({ showLogin: true });
      wrapper.find(".signup-link").simulate("click");
      expect(wrapper.state("showLogin")).toEqual(false);
    });

    it('should invoke handleLogout when logout is clicked', () => {
      wrapper.find('.logout-link').simulate('click');
      expect(mockLogoutUser).toHaveBeenCalled();
    });

    it("should invoke goToFavorites when Favorites is clicked", () => {
      wrapper.find('.favorite-link').simulate("click");
      expect(fetchUserFavorites).toHaveBeenCalled();
    });

    it("should invoke goToWatchlist when Watchlist is clicked", () => {
      wrapper.find(".watchlist-link").simulate("click");
      expect(mockDisplayWatchlist).toHaveBeenCalled();
    });

    it("should invoke resetMovies when logo is clicked", () => {
      wrapper.find(".main-heading").simulate("click");
      expect(mockResetMovies).toHaveBeenCalled();
    });

    it("should invoke resetMovies when Explore link is clicked", () => {
      wrapper.find(".explore-link").simulate("click");
      expect(mockResetMovies).toHaveBeenCalled();
    });
  });

  describe('toggleLogin', () => {
    it('should toggle the showLogin boolean', () => {
      expect(wrapper.state('showLogin')).toEqual(false);
      instance.toggleLogin();
      expect(wrapper.state("showLogin")).toEqual(true);
      instance.toggleLogin();
      expect(wrapper.state('showLogin')).toEqual(false);
    });
  });

  describe('hideLogin', () => {
    it("should set the showLogin to false", () => {
      expect(wrapper.state("showLogin")).toEqual(false);
      instance.toggleLogin();
      expect(wrapper.state("showLogin")).toEqual(true);
      instance.hideLogin();
      expect(wrapper.state("showLogin")).toEqual(false);
    });
  });

  describe("handleLogout", () => {
    it("should invoke the logoutUser dispath", () => {
      instance.handleLogout();
      expect(mockLogoutUser).toHaveBeenCalled();
    });
  });

  describe('goToFavorites', () => {
    it('should call displayFavorites with an empty array if the user has no favorites', () => {
      let noFavesUser = {
        name: 'Jacob',
        id: 1,
        favorites: []
      };
      let wrapper = shallow(
        <Nav
          user={noFavesUser}
          logoutUser={mockLogoutUser}
          resetMovies={mockResetMovies}
          displayFavorites={mockDisplayFavorites}
          displayWatchlist={mockDisplayWatchlist}
        />
      );

      wrapper.instance().goToFavorites();
      expect(mockDisplayFavorites).toHaveBeenCalledWith([]);
    });

    it('should invoke fetchUserFavorites if the user has favorited movies', () => {
      instance.goToFavorites();
      expect(fetchUserFavorites).toHaveBeenCalled();
    });

    it('should invoke cleanAllMovies after fetchUserFavorites has resolved', async () => {
      await instance.goToFavorites();
      expect(cleanAllMovies).toHaveBeenCalledWith([mockUncleanMovie]);
    });

    it('should invoke cleanAllMovies after fetchUserFavorites has resolved', async () => {
      await instance.goToFavorites();
      expect(mockDisplayFavorites).toHaveBeenCalledWith([mockCleanMovie]);
    });
  });

  describe('mapStateToProps', () => {
    it('should return an user object', () => {
      const state = { user: mockUser, test: 'test'};
      const expected = { user: state.user };
      const result = mapStateToProps(state);
      expect(result).toEqual(expected);
    });
  });

  describe("mapDispatchToProps", () => {
    it("should call dispatch when using logoutUser", () => {
      const mockDispatch = jest.fn();
      const actionToDispatch = logoutUser();
      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.logoutUser();
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });

    it("should call dispatch when using resetMovies", () => {
      const mockDispatch = jest.fn();
      const thunkToDispatch = fetchMovies('discover');
      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.resetMovies('discover');
      expect(mockDispatch).toHaveBeenCalledWith(thunkToDispatch);
    });
  });

  it("should call dispatch when using displayWatchlight", () => {
    const mockDispatch = jest.fn();
    const actionToDispatch = addWatchlistMovies(mockUser.favorites);
    const mappedProps = mapDispatchToProps(mockDispatch);
    mappedProps.displayWatchlist(mockUser.favorites);
    expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
  });
});