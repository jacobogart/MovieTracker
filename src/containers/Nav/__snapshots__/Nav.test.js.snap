// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Nav should match snapshot 1`] = `
<nav
  className="Nav"
>
  <div
    className="nav-left"
  >
    <Link
      className="main-heading"
      onClick={[Function]}
      to="/"
    >
      <h1>
        MOVIE TRACKER
      </h1>
    </Link>
    <div
      className="main-nav-wrapper"
    >
      <NavLink
        className="nav-link explore-link"
        exact={true}
        onClick={[Function]}
        to="/"
      >
        EXPLORE
      </NavLink>
      <NavLink
        className="nav-link favorite-link"
        exact={true}
        onClick={[Function]}
        to="/favorites"
      >
        FAVORITES
      </NavLink>
      <NavLink
        className="nav-link watchlist-link"
        exact={true}
        onClick={[Function]}
        to="/watchlist"
      >
        WATCHLIST
      </NavLink>
    </div>
  </div>
  <div
    className="greeting-wrapper"
  >
    <p
      className="user-greeting"
    >
      HI, 
      JACOB
      !
    </p>
    <Link
      className="logout-link"
      onClick={[Function]}
      to="/"
    >
      Logout
    </Link>
  </div>
</nav>
`;
